### PR TITLE
Replace AIR phase by using DFA to generate FUIR's clazzes

### PR DESCRIPTION
**Pull request by fridis**
The DFA is now run to incrementally create all the clazzes that are required in the IR. This avoids the overhead of the original approach that first created way to many clazzes since the air phase only performed a simple control flow analysis, and then reduce this number during DFA. 

There is one important method added: `FUIR.lookup` which is used by DFA to perform feature lookup for a call and to report that the corresponding clazz will be needed. 

This reduces the runtime for small examples using the JVM backend by about 25% and reduces the amount of code generated significantly.

### CODE REVIEW

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 1]
```diff
@@ -0,0 +1,3431 @@
...
+    _mainClazz = newClazz(mir.main().selfType())._id;
+    doesNeedCode(_mainClazz);
+    _inh = new List<>();
+    _clazzesForTypes = new TreeMap<>();
+  }
+
+
+
+  /**
+   * Clone this FUIR such that modifications can be made by optimizers.  A heir
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:257]:
```suggestion
   * Clone this FUIR such that modifications can be made by optimizers.  An heir
```

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 2]
```diff
@@ -0,0 +1,3431 @@
...
+          }
+        o = o._outer;
+      }
+
+
+
+    var t = actualType;
+
+    var cl = new Clazz(_fuiri, outerR, t, select, CLAZZ_BASE + _clazzes.size());
+    // var existing = _clazzesTM.get(cl);
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:358]: debug code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 3]
```diff
@@ -0,0 +1,3431 @@
...
+    var existing = _clazzesHM.get(cl);
+    if (existing != null)
+      {
+        result = existing;
+      }
+    else
+      {
+        /*
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:366]: debug code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 4]
```diff
@@ -0,0 +1,3431 @@
...
+              {
+                throw new Error("DUPLICATE CLAZZ: "+c+" vs "+cl+" compare is "+c.compareTo(cl));
+              }
+          }
+        */
+        result = cl;
+        _clazzes.add(cl);
+        _clazzesTM.put(cl, cl);
+        _clazzesHM.put(cl, cl);
+        /*
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:379]: debug code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 5]
```diff
@@ -0,0 +1,3431 @@
...
+      (s >= SITE_BASE,
+       s < SITE_BASE + _allCode.size(),
+       withinCode(s));
+
+    ExprKind result;
+    var e = getExpr(s);
+    if (e instanceof Clazz    )  /* Clazz represents the field we assign a value to */
+      {
+        result = ExprKind.Assign;
+        // System.out.println("###################### codeAt is "+result+" to "+e);
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:2167]: debug code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 6]
```diff
@@ -0,0 +1,3431 @@
...
+      }
+    else
+      {
+        result = exprKind(e);
+        // System.out.println("###################### codeAt is "+result+" at "+(e instanceof HasSourcePosition he ? he.pos().show() : e));
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:2172]: debug code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 7]
```diff
@@ -0,0 +1,3431 @@
...
+        res = accessedClazz(s, null);
+        if (res != null)
+          {
+            _accessedClazz.put(s, res);
+          }
+        // _accessedClazz = res; -- NYI: need Map from s to res
+      }
+    return res == null ? NO_CLAZZ : res._id;
+  }
+  //int _accessedClazz = NO_CLAZZ;
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:2414]: debug code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 8]
```diff
@@ -0,0 +1,3431 @@
...
+    var e = getExpr(s);
+    var field = switch (e)
+      {
+      case AbstractAssign a   ->
+      {
+        Clazz sClazz = clazz(a._target, outerClazz, _inh.get(s - SITE_BASE));
+        var vc = sClazz.asValue();
+        yield vc.lookup(a._assignedField);
+      }
+
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:2546]: surprising empty lines

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 9]
```diff
@@ -0,0 +1,3431 @@
...
+    int innerClazz;
+    if (accessIsDynamic(s))
+      {
+        var inner = accessedClazz(s, id2clazz(tclazz));
+        innerClazz = inner == null ? NO_CLAZZ : inner._id;
+        if (inner != null)
+          {
+            addToAccessedClazzes(s, tclazz, innerClazz);
+          }
+        /*
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:2686]: debug code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 10]
```diff
@@ -0,0 +1,3431 @@
...
+          case Routine, Intrinsic, Native, Field -> innerClazz;
+          case Abstract, Choice -> NO_CLAZZ;
+          };
+      }
+    if (innerClazz != NO_CLAZZ && codeAt(s) == ExprKind.Call)
+      {
+        doesNeedCode(innerClazz);
+      }
+
+    // System.out.println("LOOKUP for "+clazzAsString(tclazz)+" is "+clazzAsString(innerClazz)+" at "+sitePos(s).show());
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:2734]: debug code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 11]
```diff
@@ -0,0 +1,3431 @@
...
+  public void lookupDone()
+  {
+    if (CHECKS) check
+      (!_lookupDone);
+
+    _lookupDone = true;
+
+    // NYI: layout phase creates new clazzes, which is why we cannot iterate like this. Need to check why and remove this!
+    //
+    // for(var c : _clazzes)
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:2754]: orphaned code?

File: src/dev/flang/fuir/GeneratingFUIR.java [hunk 12]
```diff
@@ -0,0 +1,3431 @@
...
+import dev.flang.ast.Constant;
+import dev.flang.ast.Context;
+import dev.flang.ast.Env;
+import dev.flang.ast.Expr;
+import dev.flang.ast.FeatureName;
+import dev.flang.ast.InlineArray;
+import dev.flang.ast.NumLiteral;
+import dev.flang.ast.ResolvedNormalType;
+import dev.flang.ast.Tag;
+import dev.flang.ast.Types;
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/GeneratingFUIR.java:60]: I think the import Types should eventually be removed?

File: src/dev/flang/fuir/Clazz.java [hunk 13]
```diff
@@ -608,14 +505,14 @@ public Set<Clazz> heirs()
    */
   private Set<Clazz> directParents()
   {
-    var result = new TreeSet<Clazz>();
+    var result = new HashSet<Clazz>();
```
Comment (michaellilltokiwa) [inline src/dev/flang/fuir/Clazz.java:508]: Why Hashsets now? As far as I can see the compareTo code did not change?

File: src/dev/flang/fuir/GeneratingFUIR.java (hunk 11 above)
Comment (fridis) [reply src/dev/flang/fuir/GeneratingFUIR.java:2699]: actually not, I would like to use `var c : _clazzes`, but currently cant. will extend the NYI-comment

File: src/dev/flang/fuir/GeneratingFUIR.java (hunk 12 above)
Comment (fridis) [reply src/dev/flang/fuir/GeneratingFUIR.java:58]: All imports from `dev.flang.ast` should be replaced. by a new package that contains code common to `ast` and `fe`, so there should eventually be no more dependencies on `dev.flang.ast

File: src/dev/flang/fuir/Clazz.java (hunk 13 above)
Comment (fridis) [reply src/dev/flang/fuir/Clazz.java:508]: There was some bug in compare, but that is fixed, all `HashSet`s should be remove.d

**Review by michaellilltokiwa — APPROVED** [review]

**Pull request merged**